### DISCUSSION TITLE
Issue: Test(s) not associated (not found/executed)

### DISCUSSION THREAD

**Discussion by yapith** [fredrikaverpil/neotest-golang#161] [Configuration] [answered]
### Did you check docs and existing issues?

- [X] I have read the [documentation](https://github.com/fredrikaverpil/neotest-golang/blob/main/README.md).
- [X] I have searched the existing configuration related [discussions](https://github.com/fredrikaverpil/neotest-golang/discussions/categories/configuration).
- [X] I have updated to the latest version of Neotest.
- [X] I have updated to the latest version of neotest-golang.
- [ ] I have checked the Neotest log for errors (see README for instructions on enabling it).

### Neovim version (nvim -v)

0.10.0

### Operating system/version

Ubuntu 22.04.4 LTS

### Output from `:checkhealth neotest-golang`

```text
neotest-golang: require("neotest-golang.health").check()

Requirements ~
- OK Binary 'go' found on PATH: /snap/bin/go
- OK Found go.mod file for /home/hussein/go/src/sandbox/dir/dir_test.go in /home/hussein/go/src/sandbox/go.mod
- OK Treesitter parser for go is installed
- OK neotest is available
- OK nvim-treesitter is available
- OK nio is available
- OK plenary is available

DAP (optional) ~
- OK Binary 'dlv' found on PATH: /home/hussein/.local/share/nvim/mason/bin/dlv
- OK dap is available
- OK dapui is available
- OK dap-go is available
```

### Your Neotest and neotest-golang Lua setup

```Lua
return {
	"nvim-neotest/neotest",
	event = "VeryLazy",
	dependencies = {
		"nvim-neotest/nvim-nio",
		"nvim-lua/plenary.nvim",
		"antoinemadec/FixCursorHold.nvim",
		"nvim-treesitter/nvim-treesitter",
		{
			"fredrikaverpil/neotest-golang",
			dependencies = {
				"leoluz/nvim-dap-go",
			},
			branch = "main",
		},
	},
--	UPDATE: opts removed per fredrikaverpil's comment, it wasn't used.
	end,
	config = function()
		require("neotest").setup({
			adapters = {
				require("neotest-golang"),
			},
			-- log_level = vim.log.levels.TRACE,
			-- log = {
			-- 	enabled = true,
			--	open = true,
			-- },
			-- vim.fn.setenv("NEOTEST_LOG_FILE", vim.fn.stdpath("data") .. "/neotest.log"),
                        require("neotest.logging"):set_level(vim.log.levels.INFO)
		})
	end,
}
```

### Describe the problem

Hi, I'm testing out using the vanilla config described in the docs (with minor tweaks RE dependencies). 

Running neotest.run.run() gives me this error:
>Test(s) not associated (not found/executed):  { "/home/hussein/go/src/sandbox/dir/dir_test.go::TestFooBar" }

I'm able to run "go test -v" from the project dir just fine.

The tests summary panel seems to be able to detect the tests visually with a question mark icon next to the tests.

Curious that this doesn't output any neotest.log file to stdpath("data") per the documentation:
>require("neotest.logging"):set_level(vim.log.levels.INFO)

Also tried overriding NEOTEST_LOG_FILE, I'm still not getting neotest.log. Did I miss something here?

Hopefully I didn't miss anything silly. Love what you're doing. Thanks.

**Comment by fredrikaverpil** [comment]
Hey there,

Hm, so first off I can tell that your `opts` are not passed into your configuration function, which means they won't be used since you defined a `config` function. You have also mixed in some configuration in there which expects you to have a different neotest configuration (it looks like you may have copy-pasted from my personal configuration). Please remove the whole section `opts = function(_, opts) ... end` and let's focus on just getting the basics working without diverting from the defaults.

I've done that below, and also added in logging setup:

```lua
return {
	"nvim-neotest/neotest",
	event = "VeryLazy",
	dependencies = {
		"nvim-neotest/nvim-nio",
		"nvim-lua/plenary.nvim",
		"antoinemadec/FixCursorHold.nvim",
		"nvim-treesitter/nvim-treesitter",
		{
			"fredrikaverpil/neotest-golang",
			dependencies = {
				"leoluz/nvim-dap-go",
			},
			branch = "main",
		},
	},
	config = function()
        -- set up Neotest
		require("neotest").setup({
			adapters = {
				require("neotest-golang"),
			},
		})

        -- set up logging
        local log_level = vim.log.levels.DEBUG
        require("neotest.logging"):set_level(log_level)
        vim.notify("Logging for Neotest enabled", vim.log.levels.WARN)
        local filepath = require("neotest.logging"):get_filename()
        vim.notify("Neotest log file: " .. filepath, vim.log.levels.INFO)

	end,
}
```

This should show you both that logging is enabled and the path to the file. Do you still not see anything printed to the log file?

Do you have a separate lua configuration for Neotest?

A question mark next to the test means it hasn't run, or that it ran but there was no output from `go test` that could be correlated to the test. If this keeps occuring please have a look in the log file, and also please try to provide a minimal reproducible test example.

By the way, if you clone down this repo, you should be able to run the Go tests inside without issues. Can you do that?

EDIT: I updated the above snippet so to set up logging _after_ loading neotest.

**Reply by yapith** [reply]
Thanks for your response.

Ok looks like my machine needed rebooting (sorry pebkac). I get the neotest.log now. I don't see any errors, just WARNs and DEBUGs.

Same result running neotest.run in neotest-golang/tests/go

WARN | 2024-08-06T23:37:53Z-0700 | .../nvim/lazy/neotest-golang/lua/neotest-golang/logging.lua:74 | [neotest-golang] Test(s) not associated (not found/executed):  { "/home/hussein/go/src/sandbox/dir/dir_test.go::TestFoobaz", "/home/hussein/go/src/sandbox/dir/dir_test.go::TestFooBar" }
DEBUG | 2024-08-06T23:37:55Z-0700 | .../nvim/lazy/neotest-golang/lua/neotest-golang/logging.lua:52 | [neotest-golang] Loaded with options: {
  dap_go_opts = {},
  dev_notifications = false,
  go_test_args = { "-v", "-race", "-coverprofile=/home/hussein/go/src/sandbox/coverage.out" },
  gotestsum_args = { "--format=standard-verbose" },
  runner = "go",
  testify_enabled = false,
  warn_test_name_dupes = true,
  warn_test_not_executed = true
}


Let me know if you have any thoughts off-hand.

Thanks again.

**Reply by fredrikaverpil** [reply]
Can you provide a minimal reproducible example Go test that I can run on my machine to reproduce what you are seeing?

Did you try cloning down this very project and run the Go tests within it (using Neotest)?

**Reply by yapith** [reply]
Yes as mentioned,I cloned neotest-golang, went to neotest-golang/tests/go ran neotest.run.run(). I get the same result: Test(s) not associated (not found/executed), screenshot:

![screenshot-20240807-141435Z-selected](https://github.com/user-attachments/assets/e390a094-a9d9-4fea-8049-2d1222644790)


my neotest version: 32ff2ac
![screenshot-20240807-140657Z-selected](https://github.com/user-attachments/assets/80834743-3b79-4ee0-985a-a24be8118645)

neotest-golang version c5d557dS
![screenshot-20240807-140829Z-selected](https://github.com/user-attachments/assets/925e2a2c-1578-4858-a8c8-52e6217de437)

**Reply by yapith** [reply]
In case this helps, this is my go version: go1.22.5 linux/amd64

go env:

GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/hussein/.cache/go-build'
GOENV='/home/hussein/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/hussein/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/hussein/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/snap/go/10660'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/snap/go/10660/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.22.5'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/home/hussein/src/neotest-golang/tests/go/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--n

**Reply by fredrikaverpil** [reply]
I usually run all tests here by just starting neovim in the repo root. Then I can run the tests without issues in these files. So maybe there's some sort of issue hit when not doing that ... 🤔 

Anyway, I just cd'ed into `tests/go` and ran `nvim` in there. I still can't see what you are seeing (meaning; I cannot reproduce). 🤔 
Please see the attached screen recording. Do you think this is representative of what you are doing?

Using keymaps:

https://github.com/user-attachments/assets/3835976c-ae0c-4c00-9eb9-4029a2639967

Using `require("neotest").run.run()`:

https://github.com/user-attachments/assets/7aef945e-5a67-48ec-b09f-d418f8c2fda1

**Reply by yapith** [reply]
Yes this is exactly what I did. It's good to know that it doesn't repro for you, it's probably something off with my setup. Let me investigate some more.

Thanks again for looking into it.

**Reply by fredrikaverpil** [reply]
Sure, don't hesitate to come back here and we'll continue if you find something.

Just to help you figure out what's happening here... the adapter will run `go test` under the hood. Neotest has built its own tree of tests (you can view it in the Neotest summary window) where each test has its own "position". Think of a position as a data object holding things like the line of the test, the name of the test, full filepath to the file holding the test function etc.
In order for the neotest-golang adapter to be able to take the output from `go test` and correlate this to which "position" in the Neotest tree this belongs to, you are seeing a failure on your end for some reason.

To aid in this correlation (or association) of test output and Neotest "position" (or node in the tree, rather) I'm using `go list` to help figure out some things.

If you enable`DEBUG`-level logging you should be able to see the `go list` output and the  exact `go test` command executed. If you uncomment [this line](https://github.com/fredrikaverpil/neotest-golang/blob/c5d557df0a30e6902d53a072416045006a823762/lua/neotest-golang/process.lua#L122), you will also see the most crucial and interesting part of it all here. It should show clearly what the mismatch is, resulting in the error you are getting as it holds the data which the adapter is attempting to correlate together.

**Reply by yapith** [reply]
My repro clip:

![Peek 2024-08-07 08-01](https://github.com/user-attachments/assets/48c570d3-bdac-4215-b211-93922f9efebc)

**Reply by fredrikaverpil** [reply]
Just out of curiosity, you get this issue when you launch nvim from the `neotest-golang` repo root too, right?

**Reply by fredrikaverpil** [reply]
Btw did you get logging to work?
I think I might have made a mistake in my code example above. I think you have to call Neotest setup function and _then_ set the log level.

If you then run one single test and then attach the entire neotest log file here, it should reveal what is going on.

**Reply by fredrikaverpil** [reply]
I just enabled some more debug logging in #164. If you set the branch in your Lazy config, you'll now get the relevant output you want to inspect: `branch = "debug-logging"`. EDIT: I merged this in, so you can just update the adapter instead.

If you then run one single test and then attach the entire neotest log file here, it should reveal what is going on.

**Reply by fredrikaverpil** [reply]
I failed to mention earlier you can change test runner (this is so far [documented only in PR draft](https://github.com/fredrikaverpil/neotest-golang/pull/150) as I'm still not 100% sure how I want this to be configured). This could actually be great to try out in order to figure out what is going on as well.

Make sure you have [gotestsum](https://github.com/gotestyourself/gotestsum) installed: 
```bash
go install gotest.tools/gotestsum@latest
```

Then add the following to your neotest-golang opts: `runner = "gotestsum"`.

But first and foremost, a complete debug-level Neotest log of running a single test (or a all tests in a file, if that's when the issue is hit) would be the most insightful in figuring out the root cause of what you are experiencing.

**Reply by fredrikaverpil** [reply]
Hm. In your repro video clip, you are running all tests in the current _file_, right?

Do you also see this issue when running a single test or when running all tests in a _folder_?
Or does this only happen when running all the tests in a file?

I am really curious to take a look at what your test execution command looks like (from a log). I suspect that the package (which I get from `go list` is not proper here. 
What I really don't understand is how this differs on your machine vs my machine.

**Reply by yapith** [reply]
Yes, I have logging working. It was my fault, I had messed up symlinks and need a reboot for some reason. Apologies for the human error.

go list looks fine to me:
![screenshot-20240808-015119Z-selected](https://github.com/user-attachments/assets/210b47dc-4363-42f8-b7d7-5d32f164d1e1)

**RE Do you also see this issue when running a single test or when running all tests in a folder?**
I see this in all run test mode, single, suite, all.

Let me try gotestsum and observe the logs. I'll keep you posted.

**Comment by yapith** [comment]
[neotest.log](https://github.com/user-attachments/files/16537032/neotest.log)

Attached is the logs from a single file test of neotest-golang/tests/go/positions_test.go with gotestsum. I think you're right go list -json ./... seems to show nothing based on the attached logs: 

DEBUG | 2024-08-07T19:49:17Z-0700 | .../nvim/lazy/neotest-golang/lua/neotest-golang/logging.lua:52 | [neotest-golang] Running Go list: go list -json ./... in /home/hussein/go/src/neotest-golang/tests/go
INFO | 2024-08-07T19:49:17Z-0700 | .../nvim/lazy/neotest-golang/lua/neotest-golang/logging.lua:64 | [neotest-golang] Go list output:  

How could this line possibly returns nothing:
local output =
    vim.fn.system("cd " .. cwd .. " && " .. go_list_command_concat, " ")

which is odd because I can clearly see `go list -json ./...` output running manually:
{
	"Dir": "/home/hussein/go/src/neotest-golang/tests/go",
	"ImportPath": "github.com/fredrikaverpil/neotest-golang",
	"Name": "main",
	"Target": "/home/hussein/go/bin/neotest-golang",
	"Root": "/home/hussein/go/src/neotest-golang/tests/go",
	"Module": {
		"Path": "github.com/fredrikaverpil/neotest-golang",
		"Main": true,
		"Dir": "/home/hussein/go/src/neotest-golang/tests/go",
		"GoMod": "/home/hussein/go/src/neotest-golang/tests/go/go.mod",
		"GoVersion": "1.22.2"
	},
	"Match": [
		"./..."
	],
	"Stale": true,
	"StaleReason": "build ID mismatch",
	"GoFiles": [
		"main.go"
	],
	"Imports": [
		"fmt"
	],
	"Deps": [
		"cmp",
		"errors",
		"fmt",
		"internal/abi",
		"internal/bytealg",
		"internal/chacha8rand",
		"internal/coverage/rtcov",
		"internal/cpu",
		"internal/fmtsort",
		"internal/goarch",
		"internal/godebugs",
		"internal/goexperiment",
		"internal/goos",
		"internal/itoa",
		"internal/oserror",
		"internal/poll",
		"internal/race",
		"internal/reflectlite",
		"internal/safefilepath",
		"internal/syscall/execenv",
		"internal/syscall/unix",
		"internal/testlog",
		"internal/unsafeheader",
		"io",
		"io/fs",
		"math",
		"math/bits",
		"os",
		"path",
		"reflect",
		"runtime",
		"runtime/internal/atomic",
		"runtime/internal/math",
		"runtime/internal/sys",
		"runtime/internal/syscall",
		"slices",
		"sort",
		"strconv",
		"sync",
		"sync/atomic",
		"syscall",
		"time",
		"unicode",
		"unicode/utf8",
		"unsafe"
	],
	"TestGoFiles": [
		"positions_test.go",
		"testname_test.go"
	],
	"TestImports": [
		"testing"
	]
}
{
	"Dir": "/home/hussein/go/src/neotest-golang/tests/go/testify",
	"ImportPath": "github.com/fredrikaverpil/neotest-golang/testify",
	"Name": "testify",
	"Root": "/home/hussein/go/src/neotest-golang/tests/go",
	"Module": {
		"Path": "github.com/fredrikaverpil/neotest-golang",
		"Main": true,
		"Dir": "/home/hussein/go/src/neotest-golang/tests/go",
		"GoMod": "/home/hussein/go/src/neotest-golang/tests/go/go.mod",
		"GoVersion": "1.22.2"
	},
	"Match": [
		"./..."
	],
	"TestGoFiles": [
		"othersuite_test.go",
		"positions_test.go"
	],
	"TestImports": [
		"github.com/stretchr/testify/assert",
		"github.com/stretchr/testify/suite",
		"testing"
	]
}

**Reply by yapith** [reply, marked as answer]
Update: I figured it out the go bin installed via ubuntu snaps don't print to stdout when called from nvim, the exit code from jobstart is 0, but nothing outputed to stdout/stderr.

I had to download a go tarball and that fixed it. Thank you for your patience. Good call with the `go list`

**Comment by 39george** [comment]
Hi there,

I'm encountering the same warning message when running the entire test suite with neotest-go. The message is:

    Test(s) not associated (not found/executed)

This happens specifically when I use the run.run({ suite = true }) command. However, when I run individual test packages from the Neotest Summary using the run action, everything works as expected.

My project structure is as follows:
project_root/
```bash
├── cmd/
│   └── sitetrace/
│       └── main.go
├── config/
├── go.mod
├── go.sum
├── internal/  # Contains application packages
│   └── ... (various subdirectories)
├── tests/     # Contains integration tests
│   ├── api/
│   └── helper/
```
As you can see, I don't have any .go files directly in the project root directory. Through experimentation, I discovered that creating a dummy.go in the root allows the `test suite` command to run without warnings or errors.

Therefore, my question is: Is it mandatory to have a .go file at the root of the project for neotest-go to recognize and run the entire test suite with?

**Reply by fredrikaverpil** [reply]
Hey there @goshleg 👋 

> Is it mandatory to have a .go file at the root of the project for neotest-go to recognize and run the entire test suite with?

No, you don't need to have a .go file in the project root, or in the Neovim cwd.

You can take this very repo as an example; clone it down and open Neovim in the repo root. You should be able to successfully run the whole suite, all tests of a given file or an individual test function.

Does the test output reveal anything else, besides the "could not associate..."?

**Reply by 39george** [reply]
Without dummy.go in the project root, when I run `require("neotest").run.run({ suite = true })`, I get this warning:
> WARN | 2024-12-27T18:09:53Z+0300 | .../nvim/lazy/neotest-golang/lua/neotest-golang/logging.lua:69 | [neotest-golang] Test(s) not associated (not found/executed):  { "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/service_providers/email_client/email_client_test.go::TestEmailSendsExpectedReqPostmark", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/service_providers/client_tracking/client_tracking_test.go::TestClientTracking", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/tests/api/service_api_test.go::TestStarterApiTokenAllowsAccess", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/tests/api/service_api_test.go::TestServiceApiRoutesProtectedFromUnauthorizedAccess", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/tests/api/api_test.go::TestStarterPermissionsAllowsRestricts", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/domain/argon2/argon2_test.go::TestHashPassLightParams", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/tests/api/api_test.go::TestCreateUser", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/domain/object_key/object_key_test.go::TestFormat", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/domain/requests/service_api_test.go::TestRequestDataValidation", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/domain/object_key/object_key_test.go::TestObjectKeyParsingSuccess", '/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/domain/requests/service_api_test.go::TestRequestDataValidation::"localhost:8080"', "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/domain/object_key/object_key_test.go::TestObjectKeySerializationSuccess", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/tests/api/api_test.go::TestOpenHealthCheck", "/home/george/dev/go/projects/sitetrace/services/sitetrace_main/internal/service_providers/email_client/email_client_test.go::TestEmailSendsExpectedReqSMTPBZ" }

There are also 2497 lines with neotest logs (DEBUG), log file attached:
[neotest.log](https://github.com/user-attachments/files/18261517/neotest.log)

I cloned that repo, have run tests, and in any configuration it works correctly (with and without go files in the root), I think it could be my project only issue.

I also found this funny bug, it recognizes `gin.Engine.Run()` as test:
![Screenshot 2024-12-27 170349](https://github.com/user-attachments/assets/740580ce-04db-4bd1-a971-fa56012862c0)

**Reply by fredrikaverpil** [reply]
Sheesh, I need to look into controlling those logs better. That's too much in there... but I'll have a look in a while.

Most likely, the problem here is that treesitter finds the tests, but they are not executed cleanly due to e.g. the need of a build tag, no gcc available (because of the `go test -race`) or something else. This means that the logic that looks for output for each test cannot "associate" the test with an output.

I would've expected the adapter to pick up on any of the above mentioned crashes though. So it would be interesting to find out what's going wrong on your end.

Nice catch on that bug. It's possible that treesitter picks up on the `Run` and that we should tighten the expression to only accept `t.Run` 🤔

**Reply by fredrikaverpil** [reply]
From the log;

The `go test` command:
```
go test -json -v -coverprofile=/home/george/dev/go/projects/sitetrace/services/sitetrace_main/coverage.out sitetrace/db/...
```

Can you tell me what `sitetrace` (in `sitetrace/db/...`) is?
Is `sitetrace` the module name in your `go.mod` and do you have a `db` package?

The `go test` JSON output:

```
{
  '{"Time":"2024-12-27T18:10:58.036342201+03:00","Action":"start","Package":"sitetrace/db/sqlc"}',
  '{"Time":"2024-12-27T18:10:58.040077437+03:00","Action":"output","Package":"sitetrace/db/sqlc","Output":"\\tsitetrace/db/sqlc\\t\\tcoverage: 0.0% of statements\\n"}',
  '{"Time":"2024-12-27T18:10:58.043385905+03:00","Action":"pass","Package":"sitetrace/db/sqlc","Elapsed":0.007}'
}
```

It looks to me like no test was executed. What happens if you run the `go test` command (above) locally?

**Reply by 39george** [reply]
Thanks for looking into this. You're correct; module sitetrace is declared at the top of my root go.mod file.

Regarding the db package, I realize my previous description was inaccurate. I don't have a db package in the traditional sense. Instead, I have a db/queries directory containing SQL files and a db/sqlc package generated by sqlc. Here's a full representation of the relevant directory structure:
```
└── sitetrace_main
    ├── cmd
    │   ├── docs
    │   │   ├── docs.go
    │   │   ├── swagger.json
    │   │   └── swagger.yaml
    │   └── sitetrace
    │       └── main.go
    ├── config
    │   └── local.yaml
    ├── coverage.out
    ├── db
    │   ├── queries
    │   │   ├── api_service_queries.sql
    │   │   ├── api_token_auth_queries.sql
    │   │   ├── dashboard_queries.sql
    │   │   ├── development.sql
    │   │   ├── integration_testing_queries.sql
    │   │   ├── protected_queries.sql
    │   │   └── user_auth_queries.sql
    │   └── sqlc
    │       ├── api_service_queries.sql.go
    │       ├── api_token_auth_queries.sql.go
    │       ├── dashboard_queries.sql.go
    │       ├── db.go
    │       ├── integration_testing_queries.sql.go
    │       ├── models.go
    │       ├── protected_queries.sql.go
    │       └── user_auth_queries.sql.go
    ├── go.mod
    ├── go.sum
    ├── http_requests
    │   ├── http-client.env.json
    │   └── pack1.http
    ├── internal
    │   ├── auth
    │   │   ├── api_token_auth.go
    │   │   ├── auth.go
    │   │   └── backend.go
    │   ├── config
    │   │   └── config.go
    │   ├── domain
    │   │   ├── appstate
    │   │   │   └── appstate.go
    │   │   ├── argon2
    │   │   │   ├── argon2.go
    │   │   │   └── argon2_test.go
    │   │   ├── object_key
    │   │   │   ├── object_key.go
    │   │   │   └── object_key_test.go
    │   │   ├── requests
    │   │   │   ├── service_api.go
    │   │   │   └── service_api_test.go
    │   │   ├── user_candidate
    │   │   │   └── user_candidate.go
    │   │   └── userpass
    │   │       └── userpass.go
    │   ├── html_template
    │   │   └── html_template.go
    │   ├── httputil
    │   │   └── httputil.go
    │   ├── logger
    │   │   └── logger.go
    │   ├── middleware
    │   │   ├── api-token-auth
    │   │   │   └── api-token-auth.go
    │   │   └── user-login
    │   │       └── user-login.go
    │   ├── rabbitmqpool
    │   │   └── pool.go
    │   ├── rabbit_mq_writer
    │   │   └── rabbit_mq_writer.go
    │   ├── routes
    │   │   ├── common
    │   │   │   └── common.go
    │   │   ├── open
    │   │   │   └── open.go
    │   │   ├── protected
    │   │   │   ├── admin.go
    │   │   │   └── starter.go
    │   │   ├── service-api
    │   │   │   └── service-api.go
    │   │   └── signup
    │   │       └── signup.go
    │   ├── service_providers
    │   │   ├── captcha_verifier
    │   │   │   └── captcha_verifier.go
    │   │   ├── client_tracking
    │   │   │   ├── client_tracking.go
    │   │   │   └── client_tracking_test.go
    │   │   ├── email_client
    │   │   │   ├── email_client.go
    │   │   │   ├── email_client_test.go
    │   │   │   ├── postmark.go
    │   │   │   ├── relay.go
    │   │   │   ├── smtpbz.go
    │   │   │   └── testing.go
    │   │   ├── geoip
    │   │   │   └── geoip.go
    │   │   └── s3_storage
    │   │       └── s3_storage.go
    │   ├── startup
    │   │   ├── dbmigration
    │   │   │   └── dbmigration.go
    │   │   └── startup.go
    │   ├── tasks
    │   │   ├── post_requests.go
    │   │   └── sessions.go
    │   └── types
    │       ├── httpmethod.go
    │       └── secret_string.go
    ├── main
    ├── scripts
    │   └── swag
    ├── setenv
    ├── sqlc.yaml
    ├── tests
    │   ├── api
    │   │   ├── api_test.go
    │   │   ├── helpers.go
    │   │   └── service_api_test.go
    │   └── helper
    │       ├── common.go
    │       └── cookiejar
    │           └── cookiejar.go
    └── VERSION
```
So, the go test command you showed:

go test -json -v -coverprofile=/home/george/dev/go/projects/sitetrace/services/sitetrace_main/coverage.out sitetrace/db/...

is trying to run tests in sitetrace/db/.... Since I don't have test files directly in the db directory, and my tests are likely located elsewhere, this command won't find any tests to run. The db/sqlc package contains generated code, not test files. I think it should run `go test -json -v ./...` instead.

I was inspecting `neotest-goland/runspec/dir.lua` and in func `find_go_package_import_path` found this: if there is .go file at the root, it goes into 'perfect match' branch, and if not, it goes into 'subpackage match' branch, which in the case of this repo are:
```
All paths:
{ "github.com/fredrikaverpil/neotest-golang/x", "github.com/fredrikaverpil/neotest-golang/testify", "github.com/fredrikaverpil/neotest-golang/subpackage/subpackage2/subpackag
e3", "github.com/fredrikaverpil/neotest-golang/subpackage/subpackage2" }
Shortest path:
github.com/fredrikaverpil/neotest-golang/x
{ "go", "test", "-json", "-v", "-coverprofile=/home/george/temp/neotest-golang/tests/go/coverage.out", "github.com/fredrikaverpil/neotest-golang/..." }
```
and in case of mine are:
```
All paths:
{ "sitetrace/tests/helper/cookiejar", "sitetrace/tests/helper", "sitetrace/tests/api", "sitetrace/internal/types", "sitetrace/internal/tasks", "sitetrace/internal/startup/dbm
igration", "sitetrace/internal/startup", "sitetrace/internal/service_providers/s3_storage", "sitetrace/internal/service_providers/geoip", "sitetrace/internal/service_provider
s/email_client", "sitetrace/internal/service_providers/client_tracking", "sitetrace/internal/service_providers/captcha_verifier", "sitetrace/internal/routes/signup", "sitetra
ce/internal/routes/service-api", "sitetrace/internal/routes/protected", "sitetrace/internal/routes/open", "sitetrace/internal/routes/common", "sitetrace/internal/rabbitmqpool
", "sitetrace/internal/rabbit_mq_writer", "sitetrace/internal/middleware/user-login", "sitetrace/internal/middleware/api-token-auth", "sitetrace/internal/logger", "sitetrace/
internal/httputil", "sitetrace/internal/html_template", "sitetrace/internal/domain/userpass", "sitetrace/internal/domain/user_candidate", "sitetrace/internal/domain/requests"
, "sitetrace/internal/domain/object_key", "sitetrace/internal/domain/argon2", "sitetrace/internal/domain/appstate", "sitetrace/internal/config", "sitetrace/internal/auth", "s
itetrace/db/sqlc", "sitetrace/cmd/sitetrace", "sitetrace/cmd/docs" }
Shortest path:
sitetrace/db/sqlc
{ "go", "test", "-json", "-v", "-coverprofile=/home/george/dev/go/projects/sitetrace/services/sitetrace_main/coverage.out", "sitetrace/db/..." }
```
Interestingly, in your case, it seems to be able to correctly identify the shortest path (github.com/fredrikaverpil/neotest-golang/x) for running tests.
In my project, where there's no .go file at the root, it picks sitetrace/db/sqlc as the shortest path, which isn't the intended package for tests. Perhaps the matching logic could be improved to handle projects with different layouts?

**Reply by fredrikaverpil** [reply]
Thanks for all the great details on this issue. I believe I have a fix ready: #241

**Reply by fredrikaverpil** [reply]
I'm also addressing the `gin.Engine.Run` bug in #244

**Reply by fredrikaverpil** [reply]
@goshleg I just cut a new release 🍰 which I hope will address all of your concerns. Don't hesitate to reach out again if the problem persists on your end or if you find something else!

**Reply by 39george** [reply]
@fredrikaverpil This is fantastic! Thanks for taking the time to fix this. It's working great now!
